### PR TITLE
RSDK-1230 - Support reconfiguration for wrapper and fake arm models

### PR DESCRIPTION
**Pull request by nfranczak**
This PR adds live updates to the wrapper arm only when changing the kinematic model.
Because a wrapper arm can be used in tandem with a fake arm we add live updates for the fake arm.

### CODE REVIEW

File: components/arm/fake/fake.go
Comment (nfranczak) [inline components/arm/fake/fake.go:146]: fake arm should never reconfigure. Should this be replaced at the bottom with `config.None`?
Should the conditional at the top be changed to remove the `ok`?

Comment (raybjork) [reply components/arm/fake/fake.go:146]: You can just pull config.None out of the conditional that way it always returns config.None

File: components/arm/fake/fake.go
Comment (raybjork) [inline components/arm/fake/fake.go:131]: don't we also need to check the ArmModel? Probably is useful to refactor in such a way that the internals of the constructor can be reused here to modify the existing arm so that we don't have to repeat ourselves with writing the conditional logic to parse the config

File: components/arm/wrapper/wrapper.go
Comment (raybjork) [inline components/arm/wrapper/wrapper.go:98]: same thing here

**Review by raybjork — CHANGES_REQUESTED** [review]

File: components/arm/fake/fake.go
Comment (nfranczak) [reply components/arm/fake/fake.go:131]: sgtm

File: components/arm/fake/fake.go
Comment (nfranczak) [reply components/arm/fake/fake.go:139]: valid call out.
Spoke with ray offline about this design choice.
Please see my reply to ray's comment above for further context.

File: components/arm/fake/fake.go
Comment (nfranczak) [inline components/arm/fake/fake.go:141]: `UpdateAction` only support return value of type `config.UpdateActionType`.
Hence, if we have an error we *cannot* return it.
Proposed solution is to notify the user of the error through a debug statement. 
Open to suggestions.
LMK.

File: components/arm/wrapper/wrapper.go
Comment (nfranczak) [reply components/arm/wrapper/wrapper.go:98]: updating now.

File: components/arm/fake/fake.go
Comment (biotinker) [reply components/arm/fake/fake.go:141]: This should indeed be an error log. Logging things at an Error level do not in and of themselves have any effect on the running of the server.

File: components/arm/fake/fake.go
Comment (raybjork) [inline components/arm/fake/fake.go:103]: this can just take in the whole config as an argument right?

**Review by raybjork — CHANGES_REQUESTED** [review]

File: components/arm/fake/fake.go
Comment (spieswl) [reply components/arm/fake/fake.go:103]: Might as well, also lets you update `buildModel` without changing the signature if we decide `Fake` arms should have more attributes later.

File: components/arm/fake/fake.go
Comment (spieswl) [inline components/arm/fake/fake.go:141]: We want to be careful here that `a.model` isn't changed further down the call stack if an `error` is surfaced. Maybe worth mentioning in the docs for this function.

Comment (spieswl) [reply components/arm/fake/fake.go:141]: Specifically, `modelFromName` can possibly `nil` out your name, and result in `buildModel` returning the simplified version if the model is not supported and no path is provided. I don't think that's what a user would expect.

**Review by spieswl — CHANGES_REQUESTED** [review]
Otherwise looks pretty close, just want a little more scrutiny on the `fake` model management section and the data flow there.

File: components/arm/fake/fake.go
Comment (nfranczak) [reply components/arm/fake/fake.go]: great callout :)

File: components/arm/fake/fake.go
Comment (nfranczak) [reply components/arm/fake/fake.go]: spoke offline with William about these comments.
We came to the consensus that is a.model is nil aka err is not nil implies we fatal out to make sure the user knows they have improperly configured the fake arm.

Comment (nfranczak) [reply components/arm/fake/fake.go]: will instead use `panic`. this also closes the server and will let the user know they have improperly configured the arm.

**Review by raybjork — APPROVED** [review]

**Review by biotinker — APPROVED** [review]
LGTM, this looks really nice

**Comment by spieswl** [comment]
Just waiting for the change back to `Fatal` before I approve.

**Comment by nfranczak** [comment]
testing `fatal` is a bit tricky. but!, I got it to work :)
here are two links which really helped me out in writing the unit tests.

https://stackoverflow.com/questions/26225513/how-to-test-os-exit-scenarios-in-go

https://sr-taj.medium.com/how-to-test-methods-that-kill-your-program-in-golang-e3b879185b8a

based off what I read in these two links, to test `fatal`, we need to run the test in a subprocess.
This is why there is a test a separate test for `fatal` and another one for `fatal` with a reconfigured arm.
Totally unsure if this is the best method for how to structure the tests.
Looking for feedback.
LMK @spieswl @raybjork @biotinker

File: components/arm/fake/fake_test.go
Comment (nfranczak) [inline components/arm/fake/fake_test.go:120]: ok to `nolint` here because we did `errors.As`, right?

**Comment by nfranczak** [comment]
also, I want to call out that we **can** use `panic`. I validated through a script that this **does** cause the function to exit.
However, I do not think using `panic` is optimal. This is because it also returns a stack trace which would make parsing the error more difficult than needed. Instead, `fatal` is much more succinct.
Haha, on the other hand! testing for a panic is much simpler..

File: components/arm/fake/fake_test.go
Comment (spieswl) [inline components/arm/fake/fake_test.go:127]: Nit: TestReconfigFatalUpdate

File: components/arm/fake/fake_test.go
Comment (spieswl) [reply components/arm/fake/fake_test.go:120]: This is on a test, and its very limited in coverage. I think this is OK.

**Review by spieswl — APPROVED** [review]
LGTM

File: components/arm/fake/fake.go
Comment (edaniels) [inline components/arm/fake/fake.go:141]: So you cannot do fatal here. It's going to crash the program. UpdateAction does not support ctx/error meaning you can't *fully* use it for updating. Its actual intended usage is to check who kind of update should happen, not to process the config.

You may need to return config.Rebuild.

cc: @cheukt example of where this model isn't working too well

**Review by edaniels — CHANGES_REQUESTED** [review]

File: components/arm/wrapper/wrapper.go
Comment (edaniels) [inline components/arm/wrapper/wrapper.go:94]: better to check !ok and return early so you can dedent the rest

File: components/arm/wrapper/wrapper.go
Comment (edaniels) [inline components/arm/wrapper/wrapper.go:99]: for your debugs, use `Debugw("the msg", "error", err)`. This adds structure to the log

Comment (edaniels) [inline components/arm/wrapper/wrapper.go:99]: i'd add a comment that it's highly unlikely for this error to happen given it's checked in validate

File: components/arm/fake/fake.go
Comment (edaniels) [inline components/arm/fake/fake.go:137]: same dedent and log comments here

File: components/arm/fake/fake.go
Comment (edaniels) [inline components/arm/fake/fake.go:147]: why not rebuild?

**Review by edaniels — CHANGES_REQUESTED** [review]

File: components/arm/fake/fake.go
Comment (nfranczak) [reply components/arm/fake/fake.go:147]: agreed - I had not thought through the implications of using `config.Rebuild` instead for this case.

Let's make it explicit.
given
```
_, ok := c.ConvertedAttributes.(*AttrConfig)
```
Supposing `!ok` we return `config.Rebuild`.
We will eventually have to rebuild the fake arm through the `NewArm` method. This would then call `buildModel(cfg)` which then does:
```
armModel := cfg.ConvertedAttributes.(*AttrConfig).ArmModel
modelPath := cfg.ConvertedAttributes.(*AttrConfig).ModelFilePath
```
but since `!ok`, we would error in the `buildModel(cfg)` method. 
This is the correct way to notify the user of an improper attribute cfg.

Adding test case for this now.

**Review by raybjork — APPROVED** [review]

**Review by edaniels — APPROVED** [review]

**Review by spieswl — APPROVED** [review]
LGTM, the logging updates look much better.

**Pull request merged**
